ruma-api/src/lib.rs

```diff
@@ -254,6 +255,59 @@ pub trait OutgoingRequest {
     ) -> Result<http::Request<Vec<u8>>, IntoHttpError>;
 }
 
+/// An extension to `OutgoingRequest` which provides Appservice specific methods
+pub trait OutgoingRequestAppserviceExt: OutgoingRequest + Sized {
+    /// Tries to convert this request into an `http::Request` and appends a virtual `user_id` to
+    /// [assert Appservice identity].
+    ///
+    /// [assert Appservice identity]: https://matrix.org/docs/spec/application_service/r0.1.2#identity-assertion
+    fn try_into_http_request_with_user_id(
+        self,
+        base_url: &str,
+        access_token: Option<&str>,
+        user_id: UserId,
+    ) -> Result<http::Request<Vec<u8>>, IntoHttpError> {
+        let mut http_request = self.try_into_http_request(base_url, access_token)?;
+        let user_id_query =
+            ruma_serde::urlencoded::to_string(&[("user_id", &user_id.into_string())])?;
+
+        let uri = http_request.uri().to_owned();
+        let mut parts = uri.into_parts();
+
+        *http_request.uri_mut() = match &parts.path_and_query {
+            Some(path_and_query) => match path_and_query.query() {
+                Some(_) => {
+                    parts.path_and_query = Some(
+                        PathAndQuery::try_from(format!("{}&{}", path_and_query, user_id_query))
```

Comment:
Instead of duplicating `parts.path_and_query = Some(PathAndQuery::try_from(`, `)); parts` in every match arm, you can just return the new path+query string from the match and do everything else outside of the (outer) `match`.

Reply:
Excellent suggestion, thanks